SIPS/sip-14.md

```diff
@@ -0,0 +1,86 @@
+---
+sip: 14
+title: Dynamic Permissions
+status: Draft
+discussions-to: 
+author: Frederik Bolding (@frederikbolding)
+created: 2023-10-19
+---
+
+## Abstract
+
+This SIP proposes changes to the Snap manifest and new RPC methods that would allow Snap developers to request additional permissions dynamically at runtime. This proposal outlines some of the details around this feature.
+
+## Motivation
+
+Snaps currently have to request all permissions that they plan to use at install-time. This becomes a problem when a Snap wants to use many permissions as the installation experience suffers and the user has to either accept all permissions requested or deny the installation. This proposal provides an improvement to the experience by letting Snaps request permissions at run-time as long as those permissions are statically defined in the manifest at build-time.
+
+## Specification
+
+> Formal specifications are written in Typescript. Usage of `CAIP-N` specifications, where `N` is a number, are references to [Chain Agnostic Improvement Proposals](https://github.com/ChainAgnostic/CAIPs).
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Snap Manifest
+
+This SIP adds a new field to the Snap manifest called `dynamicPermissions`.
+This field can be used in tandem with the existing `initialPermissions`, but keep in mind that permissions in this field are not granted by installation: They MUST be requested when needed. The field follows the same format as `initialPermissions`.
+
+The new field can be specified as follows in a `snap.manifest.json` file:
+
+```json
+{
+  "initialPermissions": {
+    "endowment:transaction-insight": {}
+  },
+  "dynamicPermissions": {
+    "snap_dialog": {},
+    "snap_getBip44Entropy": [
+      {
+        "coinType": 1
+      },
+      {
+        "coinType": 3

```

Comment:
Does this mean that the Snap can never dynamically request a coinType that was not declared in the manifest? 

And what does this mean for caveats like `endowment:rpc`? Let's say dapp is `false` and snaps is `true.` Can the Snap dynamically request to make `dapp` true?

Reply:
As we have no way to merge caveats right now (and aren't planning to for v1 of this API) the limitations are:
- The snap can never request a caveat that is not predefined in the manifest
- The snap can also never request something that doesn't match exactly the definition in the manifest, you cannot even request coinType 1 and 3 separately.

Reply:
This should be mentioned in the SIP!

Reply:
Is it not already?